### PR TITLE
support local histogram

### PR DESCRIPTION
**Pull request by siddontang**
If we use metrics frequently, it may reduce the performance. For Counter and Gauge, we can simply use a int value in local and then flush to metric in batch, but it is not easy for histogram.

So we support LocalHistogram, it can be only used in single thread, and you must call `flush` interval to flush to metric. 

Benchmark 

```
test histogram::bench_histogram_local                   ... bench:           8 ns/iter (+/- 2)
test histogram::bench_histogram_local_timer             ... bench:          60 ns/iter (+/- 2)
test histogram::bench_histogram_no_labels               ... bench:          89 ns/iter (+/- 4)
test histogram::bench_histogram_timer                   ... bench:         151 ns/iter (+/- 10)
test histogram::bench_histogram_with_label_values       ... bench:         154 ns/iter (+/- 12)
```

PTAL @overvenus

### CODE REVIEW

File: src/desc.rs [hunk 1]
```diff
@@ -202,20 +206,18 @@ mod tests {
...
-            ("_9",         true ),
-            ("a9",         true ),
-            ("a_b_9_d:x_", true ),
-            ("9",          false),
-            ("9:",         false),
-            ("9_",         false),
-            ("9a",         false),
-            ("a-",         false),
-        ];
+        let tbl = [(":", true),
```
Comment (kamalmarhubi) [inline src/desc.rs]: unrelated formatting changes?

File: src/desc.rs [hunk 2]
```diff
@@ -224,21 +226,19 @@ mod tests {
...
-            (":",          false),
-            (":9",         false),
-            ("9",          false),
-            ("9:",         false),
-            ("9_",         false),
-            ("9a",         false),
-            ("a-",         false),
-            ("a_b_9_d:x_", false),
-        ];
+        let tbl = [("_", true),
```
Comment (kamalmarhubi) [inline src/desc.rs]: again

File: src/histogram.rs [hunk 3]
```diff
@@ -436,6 +443,115 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
...
+            self.histogram.core.upper_bounds.iter().enumerate().filter(|&(_, f)| v <= *f);
+        if let Some((i, _)) = iter.next() {
+            self.counts[i] += 1;
+        }
+
+        self.count += 1;
+        self.sum += v;
+    }
+
+    pub fn flush(&mut self) {
```
Comment (kamalmarhubi) [inline src/histogram.rs]: API suggestion: the usual case will be to `flush`, so make that easy. Instead of requiring user to call `flush`, make it happen automatically via `Drop`. Add a method to forget the local changes.

**Comment by kamalmarhubi** [comment]
Nice performance improvement. Is this against atomic or mutex implementation?

File: src/histogram.rs (hunk 3 above)
Comment (siddontang) [inline src/histogram.rs]: Hi @kamalmarhubi 

In our program, we will keep the local histogram and flush it interval, so `flush` is needed for us and I think we also need `drop` too, I will add it.

**Comment by siddontang** [comment]
Hi @kamalmarhubi 

> Is this against atomic or mutex implementation

No, it is just a improvement for performance only. Most of the cases, we will use Histogram, but in some cases, especially in TiKV, we meet performance reduction when using mutex `*Vec` metrics too frequently, so we first cache these in local and flush interval, like:

```
// for counter
local c: f64 = 0;
// add counter frequently
for in 1..100000 {
    c += 1;
}
// finally
COUNTER_VEC.with_label_values(&["a"]).inc_by(c);
```

File: src/histogram.rs (hunk 3 above)
Comment (kamalmarhubi) [inline src/histogram.rs]: I think it'd be good to add a method like

```
fn forget(&mut self) {
    *self = LocalHistogramCore::new(self.histogram);
}
```

to allow removing any pending changes so they don't get applied via the `Drop` impl.

Comment (kamalmarhubi) [inline src/histogram.rs]: (Not sure of the best name for it.)

Comment (siddontang) [inline src/histogram.rs]: I prefer `clear` :-)

**Comment by siddontang** [comment]
PTAL @kamalmarhubi

File: src/histogram.rs [hunk 4]
```diff
@@ -436,6 +443,140 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
...
+        let core = LocalHistogramCore::new(histogram);
+        LocalHistogram { core: Rc::new(RefCell::new(core)) }
+    }
+
+    /// `observe` adds a single observation to the `Histogram`.
+    pub fn observe(&self, v: f64) {
+        self.core.borrow_mut().observe(v);
+    }
+
+    /// `start_timer`
```
Comment (overvenus) [inline src/histogram.rs]: Please add some documents.

File: src/histogram.rs [hunk 5]
```diff
@@ -436,6 +443,140 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
...
+        if let Some((i, _)) = iter.next() {
+            self.counts[i] += 1;
+        }
+
+        self.count += 1;
+        self.sum += v;
+    }
+
+    pub fn clear(&mut self) {
+        for v in self.counts.iter_mut() {
```
Comment (overvenus) [inline src/histogram.rs]: How about `for v in &mut self.counts { ... }`

File: src/histogram.rs [hunk 6]
```diff
@@ -436,6 +443,140 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
...
+        self.sum = 0.0;
+    }
+
+    pub fn flush(&mut self) {
+        // No cached metric, return.
+        if self.count == 0 {
+            return;
+        }
+
+        let h = self.histogram.clone();
```
Comment (overvenus) [inline src/histogram.rs]: Actually we can avoid `clone()`.

```rust
        {
            let h = &mut self.histogram;
            // ...
        }
```

File: src/histogram.rs [hunk 7]
```diff
@@ -587,4 +728,35 @@ mod tests {
...
+        let opts = HistogramOpts::new("test_histogram_local", "test histogram local help")
+            .buckets(buckets.clone());
+        let histogram = Histogram::with_opts(opts).unwrap();
+        let local = histogram.local();
+
+        let check = |count, sum| {
+            let m = histogram.metric();
+            let proto_histogram = m.get_histogram();
+            assert_eq!(proto_histogram.get_sample_count(), count);
+            assert_eq!(proto_histogram.get_sample_sum(), sum);
```
Comment (overvenus) [inline src/histogram.rs]: Clippy blames: 

```
warning: strict comparison of f32 or f64, #[warn(float_cmp)] on by default
   --> src/histogram.rs:746:13
    |
746 |             assert_eq!(proto_histogram.get_sample_sum(), sum);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
help: consider comparing them within some error
    | if ! (* left_val - * right_val).abs() < error {
note: std::f32::EPSILON and std::f64::EPSILON are available.
```

**Comment by siddontang** [comment]
PTAL @disksing

File: src/histogram.rs [hunk 8]
```diff
@@ -306,6 +308,11 @@ impl Histogram {
     pub fn start_timer(&self) -> HistogramTimer {
         HistogramTimer::new(self.clone())
     }
+
+    /// `local` returns a `LocalHistogram` for single thread usage.
+    pub fn local(&self) -> LocalHistogram {
+        LocalHistogram::new(self.clone())
+    }
```
Comment (disksing) [inline src/histogram.rs]: Will it be a problem if other thread keeps using the old `Histogram` after creating the `LocalHistogram`?

File: src/histogram.rs [hunk 9]
```diff
@@ -436,6 +443,142 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
...
+pub struct LocalHistogramCore {
+    histogram: Histogram,
+    counts: Vec<u64>,
+    count: u64,
+    sum: f64,
+}
+
+/// `LocalHistogram` is used for performance and in single-thread.
+/// Sometimes, if you very care the performance, you can use the `LocalHistogram`
+// and then flush the metric interval.
```
Comment (disksing) [inline src/histogram.rs]: // ->  ///

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @overvenus @kamalmarhubi

**Comment by kamalmarhubi** [comment]
LGTM
